woocommerce/woocommerce-ios
# Image Upload: observe the same image statuses in Edit Product and Edit Images screens

"Update UX" part 1 for #1713 

## Background

First I apologized for the long PR, about 350 new diffs and 69 deleted diffs are for updating the unit tests 🙇🏻‍♀️Before this PR, we only have one screen (`ProductImagesViewController`) handling the Product image upload states. This PR implements parts of the UX as in p99K0U-2mN-p2#comment-4732:
- [Implemented in this PR] If the user taps “Done” while some images are still being uploaded, we will keep track of the images that are in-progress, display the in-progress UI (blurred image + circle spinner) for these images on the Edit Product screen, and then update the UI for these images once they are uploaded.
  - [TODO] When the user taps “Done” on the Edit Product screen, we first show the in-progress UI, wait for all images to be uploaded and then update the Product remotely.
  - [TODO] When the user navigates away from the Edit Product screen, no changes are made to the Product and these images keep being uploaded in the background.
- [TODO] If the user navigates away from the Edit Product Images screen while some images are still being uploaded, these images keep being uploaded in the background but they won’t be added to the Product.

## Changes

NOTE: I'm totally open to any other implementation ideas or concerns! I've changed the implementation like 2 times and don't mind more rewrites to make the code more clear 😛  

- Created `ProductImagesService` that allows anyone to observe its read-only `productImageStatuses: [ProductImageStatus]` as the source of truth in an Edit Product session. `ProductImagesService` manages the array of `ProductImageStatus` with 2 actions for now:
  - An image is deleted by the user: `deleteProductImage(_ productImage: ProductImage)` is called
  - An image asset is selected by the user to upload to WP media library: `uploadMediaAssetToSiteMediaLibrary(asset: PHAsset)` is called
  - The read-only `productImageStatuses` state can be observed by multiple observers via an `ObservationToken` that can be cancelled or automatically removed when the observer is deallocated - I borrowed the idea from https://www.swiftbysundell.com/articles/observers-in-swift-part-2/ and any feedback are welcome!
- Created `ProductImagesProvider` that provides a `UIImage` given a `PHAsset` or `ProductImage`, and an update function when a `PHAsset` has been uploaded to a `ProductImage` for caching implementation
  - Created `DefaultProductImagesProvider` that implements `ProductImagesProvider`
- Added `ProductImagesService` and `ProductImagesProvider` to `ProductFormViewController` (Edit Product screen), and all the classes within the same Edit Product session share these two instances so that the images are cached in `DefaultProductImagesProvider` and the product image statuses state remains the same
- Replaced existing code that fetches an `UIImage` given a `ProductImage` or `PHAsset` with `ProductImagesProvider`
- Updated the Edit Product images header cell (`ProductImagesHeaderTableViewCell`) to use the same data `productImageStatuses` from `ProductImagesService`
- Added/updated unit tests:
  - Moved `MockImageCache` and `MockImageDownloader` from `DefaultImageServiceTests` for reuse
  - Added unit tests for `ProductImagesService` and `DefaultProductImagesProvider`

## Testing

- Go to the Products tab
- Tap on a simple Product
- Tap on the "+" cell on the images header row to edit images
- Tap on an existing image, and tap the trash icon to delete it
- Tap "Add Photos" to add some photos via Photo Library or camera --> the images should have the cells with spinner
- While the images are being uploaded, tap "Done" --> the same in-progress cells should be shown on the Edit Product screen with the image deleted previously gone
- Tap on the "+" cell on the images header row to edit images again --> the same in-progress cells should be shown on the Edit Product screen
- Wait for the images to be uploaded --> the cells should transition from the in-progress state to images after they are uploaded successfully
- Tap "Done" to go back to Edit Product --> the same images should be displayed on the images scroll view in the header

## Example screenshots

Edit Product Images | Edit Product
-- | --
![Simulator Screen Shot - iPhone 11 - 2020-02-20 at 15 51 51](https://user-images.githubusercontent.com/1945542/74912123-fa075200-53f8-11ea-9b31-1fc3de3f05e8.png) | ![Simulator Screen Shot - iPhone 11 - 2020-02-20 at 15 51 54](https://user-images.githubusercontent.com/1945542/74912134-ff649c80-53f8-11ea-8d09-58369041aaee.png)


Update release notes:
